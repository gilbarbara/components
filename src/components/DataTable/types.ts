import { CSSProperties, MouseEventHandler, ReactNode } from 'react';
import { Simplify, StringOrNull, StringOrNumber } from '@gilbarbara/types';

import { LoaderType } from '~/components/Loader/types';

import {
  OmitElementProps,
  SortDirection,
  StyledProps,
  VariantWithTones,
  WithAccent,
  WithFlexItem,
  WithLayout,
  WithMargin,
  WithPadding,
  WithRadius,
} from '~/types';

export interface DataTableColumn<T = string> {
  /**
   * Disable sorting for the column.
   * @default false
   */
  disableSort?: boolean;
  /**
   * Hide the column on smaller screens.
   * @default false
   */
  hideOnResponsive?: boolean;
  /**
   * Set the column as an action column and disable sorting.
   * @default false
   */
  isAction?: boolean;
  /**
   * The key to use for the column.
   */
  key: T;
  /**
   * The maximum width for the column.
   */
  max?: StringOrNumber;
  /**
   * The minimum width for the column.
   */
  min?: StringOrNumber;
  /**
   * The fixed width for the column.
   * Overrides `min` and `max`.
   */
  size?: StringOrNumber;
  /**
   * The title for the column.
   */
  title: ReactNode;
}

export type DataTableRowContent = ReactNode | { label: ReactNode; value: string };

export type DataTableRow<T extends string> = Simplify<
  Record<T, DataTableRowContent> & { id?: StringOrNumber }
>;

export interface DataTableKnownProps<T extends string>
  extends StyledProps,
    WithAccent,
    WithFlexItem,
    WithLayout,
    WithMargin,
    WithPadding,
    WithRadius {
  /**
   * Component background color
   * If not set, the component will use 'gray.800' for darkMode otherwise 'gray.50'
   */
  bg?: VariantWithTones;
  /**
   * The breakpoint to switch to responsive mode.
   * @default 768 */
  breakpoint?: number;
  /**
   * Remove background and padding
   * @default false
   */
  clean?: boolean;
  /**
   * The columns to display
   */
  columns: Array<DataTableColumn<T>>;
  /**
   * The data to display
   */
  data: Array<DataTableRow<T>>;
  /**
   * The default column to sort by
   */
  defaultSortColumn?: T;
  /**
   * The default sort direction
   * @default asc
   */
  defaultSortDirection?: SortDirection;
  /**
   * Disable scroll to top on page change
   * @default false
   */
  disableScroll?: boolean;
  /**
   * The size of the loader.
   * @default 128
   */
  loaderSize?: number;
  /**
   * The type of loader to display.
   */
  loaderType?: LoaderType;
  /**
   * Show a loader while data is loading.
   * @default false
   */
  loading?: boolean;
  /**
   * The maximum number of rows to display per page.
   * @default 10 */
  maxRows?: number;
  /**
   * Custom message when no results are found.
   * @default Nothing found
   */
  noResults?: ReactNode;
  onClickPage?: (page: number, totalPages: number) => void;
  onClickSort?: (sortedBy: T, sortDirection: SortDirection) => void;
  /**
   * Show pagination
   * @default true
   */
  pagination?: boolean;
  /**
   * Use remote paginated data.
   */
  remote?: {
    currentPage: number;
    totalPages: number;
    useInternalSorting?: boolean;
  };
  /**
   * Rearrange columns on smaller screens.
   * @default false
   */
  responsive?: boolean;
  /**
   * The duration of the scroll animation on page change.
   * @default 400
   */
  scrollDuration?: number;
  /**
   * The element to scroll to on page change.
   */
  scrollElement?: HTMLElement | null;
  /**
   * The margin to apply when scrolling to the element.
   */
  scrollMargin?: number;
  /**
   * Set the header sticky.
   * @default false
   */
  stickyHeader?: boolean;
  style?: CSSProperties;
  /**
   * The width of the table.
   * @default window.innerWidth */
  width?: number;
}

export type DataTableProps<T extends string = string> = Simplify<
  OmitElementProps<HTMLDivElement, DataTableKnownProps<T>, 'data' | 'wrap'>
>;

export interface DataTableHeadProps extends WithAccent, Pick<DataTableProps, 'clean' | 'columns'> {
  darkMode: boolean;
  isDisabled: boolean;
  isResponsive: boolean;
  onClick: MouseEventHandler;
  sortBy: StringOrNull;
  sortDirection: SortDirection;
  stickyHeader: boolean;
}

export interface DataTableBodyProps
  extends WithAccent,
    Pick<DataTableProps, 'clean' | 'columns' | 'data' | 'loaderSize' | 'loaderType' | 'loading'> {
  darkMode: boolean;
  isResponsive: boolean;
  sortColumn?: string;
}
