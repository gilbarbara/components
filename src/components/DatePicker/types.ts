import { ReactNode } from 'react';
import { DayPickerBase, DayPickerRangeProps, DayPickerSingleProps } from 'react-day-picker';
import { Simplify, StringOrNumber } from '@gilbarbara/types';

import {
  Alignment,
  Color,
  StyledProps,
  WithAccent,
  WithBorder,
  WithBorderless,
  WithDisabled,
  WithMargin,
  WithOpen,
  WithPadding,
  WithRadius,
  WithShadow,
} from '~/types';

export type DatePickerRangeParameter = [from?: string, to?: string];

export type DatePickerRangeClickHandler = (range: DatePickerRangeParameter) => void;
export type DatePickerSingleClickHandler = (isoDate: string) => void;

export interface DatePickerBaseProps extends StyledProps, WithAccent<Color> {
  /**
   * @default Go to today
   */
  currentMonthLabel?: ReactNode;
  fromDate?: string | Date;
  toDate?: string | Date;
}

export interface DatePickerLayoutProps
  extends WithBorder,
    WithMargin,
    WithPadding,
    WithRadius,
    WithShadow {}

export interface DatePickerRangeKnownProps
  extends DatePickerBaseProps,
    DatePickerLayoutProps,
    Omit<DayPickerRangeProps, 'fromDate' | 'mode' | 'selected' | 'onSelect' | 'toDate'> {
  /**
   * The date format to use when displaying the dates.
   * @default en-US
   */
  formatLocale?: string;
  /**
   * For internal use with DatePickerInput
   * @private
   */
  onApply?: (selected: DatePickerRangeParameter) => void;
  onChange?: DatePickerRangeClickHandler;
  readOnly?: boolean;
  /**
   * The initial date.
   */
  selected?: DatePickerRangeParameter;
  /**
   * For internal use with DatePickerInput
   * @private
   */
  showApply?: boolean;
}

export type DatePickerRangeProps = Simplify<DatePickerRangeKnownProps>;

export interface DatePickerSingleKnownProps
  extends DatePickerBaseProps,
    DatePickerLayoutProps,
    Omit<DayPickerSingleProps, 'fromDate' | 'mode' | 'onSelect' | 'selected' | 'toDate'> {
  onChange?: DatePickerSingleClickHandler;
  readOnly?: boolean;
  /**
   * The initial date.
   */
  selected?: string;
}

export type DatePickerSingleProps = Simplify<DatePickerSingleKnownProps>;

export interface DatePickerSelectorBaseProps
  extends DatePickerBaseProps,
    Omit<DayPickerBase, 'disabled' | 'fromDate' | 'mode' | 'selected' | 'toDate'>,
    WithBorderless,
    WithDisabled,
    WithOpen,
    WithMargin {
  /**
   * The date format to use when displaying the dates.
   * @default en-US
   */
  formatLocale?: string;
  large?: boolean;
  /**
   * Whether to show the calendar in range or single mode.
   * @default single
   */
  mode?: 'range' | 'single';
  /**
   * Add a hidden input with the given name.
   */
  name?: string;
  /**
   * For internal use with DatePickerSelector
   * @private
   */
  onApply?: (selected: DatePickerRangeParameter) => void;
  placeholder?: string;
  /**
   * The position of the calendar relative to the input.
   * @default right
   */
  position?: Alignment;
  /**
   * The initial date.
   */
  selected?: DatePickerRangeParameter | string;
  /**
   * The separator for date ranges.
   * @default ' â€” '
   */
  separator?: string;
  /**
   * Show a button to apply the selected range.
   * @default false
   */
  showRangeApply?: boolean;
  /**
   * The width of the selector.
   * It has a min-width of 200px.
   */
  width?: StringOrNumber;
}

export type DatePickerSelectorProps = Simplify<
  DatePickerSelectorBaseProps &
    (
      | {
          mode: 'range';
          onChange?: DatePickerRangeClickHandler;
        }
      | {
          mode: 'single';
          onChange?: DatePickerSingleClickHandler;
        }
    )
>;
