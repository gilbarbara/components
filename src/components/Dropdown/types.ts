import { HiddenInput, Option, Props } from '@gilbarbara/react-dropdown';
import { Simplify, StringOrNumber } from '@gilbarbara/types';

import { StyledProps, WithAccent, WithBorderless, WithMargin, WithOpen } from '~/types';

export type DropdownBaseProps = Omit<
  Props,
  | 'className'
  | 'clearComponent'
  | 'clearOnSelect'
  | 'contentComponent'
  | 'create'
  | 'handleComponent'
  | 'inputComponent'
  | 'loadingComponent'
  | 'menuComponent'
  | 'menuItemComponent'
  | 'noDataComponent'
  | 'onClearAll'
  | 'onCreate'
  | 'onSelectAll'
  | 'optionComponent'
  | 'options'
  | 'searchFn'
  | 'secondaryPlaceholder'
  | 'separatorComponent'
  | 'style'
  | 'styles'
>;

export interface DropdownKnownProps
  extends StyledProps,
    WithAccent,
    WithBorderless,
    WithMargin,
    WithOpen,
    DropdownBaseProps {
  /** @default false */
  allowCreate?: boolean;
  /** @default false */
  closeMultiOnSelect?: boolean;
  /**
   * If set, an input with type hidden will be added to the component with the value of the selected option(s).
   * In case of multiple items, the value will be a string concatenated with "separator".
   */
  inputOptions?: HiddenInput;
  items: Option[];
  large?: boolean;
  /** @default 260 */
  menuMaxHeight?: number;
  onClear?: () => void;
  onCreate?: (item: string, close: () => void) => void;
  onSearch?: (value: string) => void;
  /** @default 260 */
  width?: StringOrNumber;
}

export type DropdownProps = Simplify<DropdownKnownProps>;
