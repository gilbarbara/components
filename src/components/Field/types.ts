import { ChangeEventHandler, FocusEventHandler, ReactNode } from 'react';
import { FieldValues, UseFormGetValues } from 'react-hook-form';
import { Simplify, StringOrNumber } from '@gilbarbara/types';

import { DatePickerRangeParameter } from '~/components/DatePicker/types';

import {
  CheckboxItem,
  DropdownOption,
  InputTypes,
  RadioItem,
  ValidatePasswordOptions,
  WithAccent,
  WithBorderless,
  WithDisabled,
} from '~/types';
import type { DatePickerSelectorProps, DropdownProps, FormGroupProps } from '~/types/props';

export type RegisterOptionsProps = Simplify<
  FieldBaseProps & {
    getValues: UseFormGetValues<FieldValues>;
  } & (
      | {
          formatter?: FieldInputProps['formatter'];
          type: Exclude<FieldTypes, 'password'>;
          validationOptions?: never;
        }
      | {
          formatter?: never;
          type: 'password';
          validationOptions?: ValidatePasswordOptions;
        }
    )
>;

export type FieldTypes =
  | InputTypes
  | 'checkbox'
  | 'datePicker'
  | 'dropdown'
  | 'radio'
  | 'select'
  | 'toggle'
  | 'textarea';

export type FieldValidations = 'email' | `equalsTo:${string}` | 'password' | 'phoneBR' | 'phoneUS';

export interface FieldInputHandlers<
  T extends HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement,
> {
  onBlur?: FocusEventHandler<T>;
  onChange?: ChangeEventHandler<T>;
  onFocus?: FocusEventHandler<T>;
}

interface FieldExcludedProps {
  children?: never;
  datePickerProps?: never;
  dropdownProps?: never;
  formatter?: never;
  items?: never;
  onBlur?: never;
  onFocus?: never;
  validationOptions?: never;
}

export interface FieldBaseProps
  extends Pick<
      FormGroupProps,
      'assistiveText' | 'hideAssistiveText' | 'inline' | 'label' | 'required' | 'style'
    >,
    WithAccent,
    WithBorderless,
    WithDisabled {
  autoComplete?: string;
  clearError?: () => void;
  debug?: boolean;
  id?: string;
  maxLength?: number;
  minLength?: number;
  name: string;
  placeholder?: string;
  readOnly?: boolean;
  setValueAs?: (value: any) => any;
  skipValidation?: boolean;
  validations?: FieldValidations[];
  value?: any;
}

export type FieldCheckboxProps = Simplify<
  FieldBaseProps &
    Omit<FieldExcludedProps, 'items'> & {
      items: CheckboxItem[];
      onChange?: (value: Array<string>) => void;
      type: 'checkbox';
    }
>;

export type FieldDatePickerProps = Simplify<
  FieldBaseProps &
    Omit<FieldExcludedProps, 'datePickerProps'> & {
      datePickerProps?: Simplify<Omit<DatePickerSelectorProps, 'placeholder'>>;
      onChange?: (selection: DatePickerRangeParameter | string) => void;
      type: 'datePicker';
    }
>;

export type FieldDropdownProps = Simplify<
  FieldBaseProps &
    Omit<FieldExcludedProps, 'dropdownProps' | 'items'> & {
      dropdownProps?: Omit<
        DropdownProps,
        'disabled' | 'inputOptions' | 'items' | 'onChange' | 'placeholder' | 'width'
      >;
      items: DropdownOption[];
      onChange?: (value: StringOrNumber | Array<StringOrNumber>) => void;
      type: 'dropdown';
    }
>;

export type FieldInputProps = Simplify<
  FieldBaseProps &
    Omit<FieldExcludedProps, 'formatter' | 'onBlur' | 'onFocus'> &
    Omit<FieldInputHandlers<HTMLInputElement>, 'onChange'> & {
      formatter?: 'money' | 'number' | 'phoneBR' | 'phoneUS';
      onChange?: (value: string) => void;
      type: Exclude<InputTypes, 'password'>;
    }
>;

export type FieldPasswordProps = Simplify<
  FieldBaseProps &
    Omit<FieldExcludedProps, 'onBlur' | 'onFocus' | 'validationOptions'> &
    Omit<FieldInputHandlers<HTMLInputElement>, 'onChange'> & {
      onChange?: (value: string) => void;
      type: 'password';
      validationOptions?: ValidatePasswordOptions;
    }
>;

export type FieldRadioProps = Simplify<
  FieldBaseProps &
    Omit<FieldExcludedProps, 'items'> & {
      items: RadioItem[];
      onChange?: (value: string) => void;
      type: 'radio';
    }
>;

export type FieldSelectProps = Simplify<
  FieldBaseProps &
    Omit<FieldExcludedProps, 'children' | 'onBlur' | 'onFocus'> &
    Omit<FieldInputHandlers<HTMLSelectElement>, 'onChange'> & {
      children: ReactNode[];
      onChange?: (value: string) => void;
      type: 'select';
    }
>;

export type FieldTextareaProps = Simplify<
  FieldBaseProps &
    Omit<FieldExcludedProps, 'onBlur' | 'onFocus'> &
    Omit<FieldInputHandlers<HTMLTextAreaElement>, 'onChange'> & {
      onChange?: (value: string) => void;
      rows?: number;
      type: 'textarea';
    }
>;

export type FieldToggleProps = Simplify<
  FieldBaseProps &
    FieldExcludedProps & {
      onChange?: (value: boolean) => void;
      type: 'toggle';
    }
>;

export type FieldProps =
  | FieldCheckboxProps
  | FieldDatePickerProps
  | FieldDropdownProps
  | FieldInputProps
  | FieldPasswordProps
  | FieldRadioProps
  | FieldSelectProps
  | FieldTextareaProps
  | FieldToggleProps;
