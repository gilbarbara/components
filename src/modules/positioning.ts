import { CSSObject } from '@emotion/react';

import { parseFloatNumber } from '~/modules/helpers';

import { FloatingAlignment, FloatingPlacement, Position } from '~/types';

interface ElementDimensions {
  height: number;
  width: number;
}

export interface GetComputedPlacementOptions {
  preferredPlacements?: FloatingPlacement[];
}

export interface GetElementDimensionsResult {
  height: number;
  width: number;
}

export interface GetFloatingStylesOptions {
  distance?: number;
  offset?: number;
}

export const positionToDiretionMap = {
  auto: 'down',
  top: 'up',
  bottom: 'down',
  left: 'left',
  right: 'right',
} as const;

export function getComputedPlacement(
  placement: FloatingPlacement,
  triggerElement: HTMLElement | null,
  contentElement: HTMLElement | null,
  options: GetComputedPlacementOptions = {},
): FloatingPlacement {
  const { innerHeight: height, innerWidth: width } = window;
  const { preferredPlacements } = options;

  if (!triggerElement || !contentElement) {
    return placement === 'auto' ? 'bottom' : placement;
  }

  const triggerRect = triggerElement.getBoundingClientRect();
  const { height: contentHeight, width: contentWidth } = getElementDimensions(contentElement);

  const spaces = {
    top: triggerRect.top,
    bottom: height - triggerRect.bottom,
    left: triggerRect.left,
    right: width - triggerRect.right,
  };

  let computedPosition: Position;
  let computedAlignment: FloatingAlignment | undefined;

  if (placement === 'auto') {
    // Get all positions with sufficient space
    const positionsWithSpace = Object.entries(spaces).filter(
      ([_, availableSpace]) => availableSpace >= contentHeight,
    );

    // Check if 'bottom' is among the positions with sufficient space
    const bottomSpace = positionsWithSpace.find(([pos]) => pos === 'bottom');

    if (bottomSpace) {
      // If 'bottom' has enough space, use it
      computedPosition = 'bottom';
    } else {
      // Otherwise, sort remaining positions by available space and pick the best one
      const bestFit = positionsWithSpace
        .sort((a, b) => b[1] - a[1])
        .map(([pos]) => pos as Position);

      // Still default to 'bottom' if nothing works
      computedPosition = bestFit[0] || 'bottom';
    }
  } else {
    [computedPosition, computedAlignment] = placement.split('-') as [
      Position,
      FloatingAlignment | undefined,
    ];
  }

  // Check for horizontal overflow considering the full placement
  const wouldOverflowHorizontally = (pos: Position, alignment?: FloatingAlignment) => {
    let horizontalPosition: number;

    if (['left', 'right'].includes(pos)) {
      // For left/right placements, set the horizontal position
      horizontalPosition = pos === 'left' ? triggerRect.left - contentWidth : triggerRect.right;
    } else if (alignment) {
      horizontalPosition =
        alignment === 'start' ? triggerRect.left : triggerRect.right - contentWidth;
    } else {
      horizontalPosition = triggerRect.left + (triggerRect.width - contentWidth) / 2;
    }

    return horizontalPosition < 0 || horizontalPosition + contentWidth > width;
  };

  // Check for vertical overflow considering the full placement
  const wouldOverflowVertically = (pos: Position, alignment?: FloatingAlignment) => {
    let verticalPosition: number;

    if (['bottom', 'top'].includes(pos)) {
      // For top/bottom placements, set the vertical position
      verticalPosition = pos === 'top' ? triggerRect.top - contentHeight : triggerRect.bottom;
    } else if (alignment) {
      verticalPosition =
        alignment === 'start' ? triggerRect.top : triggerRect.bottom - contentHeight;
    } else {
      verticalPosition = triggerRect.top + (triggerRect.height - contentHeight) / 2;
    }

    return verticalPosition < 0 || verticalPosition + contentHeight > height;
  };

  // Check if the current computed placement would overflow
  if (
    !wouldOverflowHorizontally(computedPosition, computedAlignment) &&
    !wouldOverflowVertically(computedPosition, computedAlignment)
  ) {
    return `${computedPosition}${computedAlignment ? `-${computedAlignment}` : ''}` as FloatingPlacement;
  }

  const defaultPlacementsOrder =
    {
      auto: [
        'bottom',
        'bottom-start',
        'bottom-end',
        'top',
        'top-start',
        'top-end',
        'right',
        'right-start',
        'right-end',
        'left',
        'left-start',
        'left-end',
      ],
      bottom: ['bottom-end', 'bottom-start', 'top', 'top-start', 'top-end', 'right', 'left'],
      left: [
        'left-start',
        'left-end',
        'top-start',
        'top',
        'top-end',
        'bottom-start',
        'bottom',
        'bottom-end',
        'right',
      ],
      right: [
        'right-start',
        'right-end',
        'bottom-end',
        'bottom',
        'bottom-start',
        'top-end',
        'top',
        'top-start',
        'left',
      ],
      top: ['top-end', 'top-start', 'bottom', 'bottom-start', 'bottom-end', 'right', 'left'],
    }[computedPosition] || [];

  // Define fallback order based on the requested placement
  const placementsOrder = preferredPlacements ?? defaultPlacementsOrder;

  // Find the first available placement with the least overflow
  let lowestOverflow = Infinity;
  let bestPlacement: FloatingPlacement | null = null;

  for (const p of placementsOrder) {
    const [pos, align] = p.split('-') as [Position, FloatingAlignment | undefined];

    // Calculate overflow scores
    const horizontalOverflowAmount = wouldOverflowHorizontally(pos, align) ? 1 : 0;
    const verticalOverflowAmount = wouldOverflowVertically(pos, align) ? 1 : 0;
    const totalOverflow = horizontalOverflowAmount + verticalOverflowAmount;

    // Check if this placement has space on its primary axis
    const hasSpaceOnPrimaryAxis =
      pos === 'top' || pos === 'bottom'
        ? spaces[pos] >= contentHeight
        : spaces[pos] >= contentWidth;

    if (hasSpaceOnPrimaryAxis && totalOverflow < lowestOverflow) {
      lowestOverflow = totalOverflow;
      bestPlacement = p as FloatingPlacement;

      // If we found a placement with no overflow, use it immediately
      if (totalOverflow === 0) {
        break;
      }
    }
  }

  // If we found a best placement without an alignment but the original had one,
  // try to preserve the original alignment if possible
  if (bestPlacement && !bestPlacement.includes('-') && computedAlignment) {
    const candidatePlacement = `${bestPlacement}-${computedAlignment}` as FloatingPlacement;
    const [pos, align] = candidatePlacement.split('-') as [Position, FloatingAlignment];

    if (!wouldOverflowHorizontally(pos, align) && !wouldOverflowVertically(pos, align)) {
      return candidatePlacement;
    }
  }

  return bestPlacement ?? 'bottom';
}

/**
 * Gets the dimensions of an element with its margins
 */
export function getElementDimensions(contentElement: HTMLElement): ElementDimensions {
  const contentStyles = getComputedStyle(contentElement);

  const margin = Math.round(parseFloatNumber(contentStyles.margin));
  const marginX = Math.round(
    parseFloatNumber(contentStyles.marginLeft) + parseFloatNumber(contentStyles.marginRight),
  );
  const marginY = Math.round(
    parseFloatNumber(contentStyles.marginTop) + parseFloatNumber(contentStyles.marginBottom),
  );

  // Get dimensions from computed style and scroll dimensions
  const styleHeight = Math.round(parseFloatNumber(contentStyles.height));
  const styleWidth = Math.round(parseFloatNumber(contentStyles.width));
  const scrollHeight = Math.round(contentElement.scrollHeight);
  const scrollWidth = Math.round(contentElement.scrollWidth);

  // Use the largest value between style dimensions and scroll dimensions
  const height = Math.max(styleHeight, scrollHeight) + (marginY || margin);
  const width = Math.max(styleWidth, scrollWidth) + (marginX || margin);

  return { height, width };
}

/**
 * Returns computed styles for a floating element based on placement and distance.
 */
export function getFloatingStyles(
  placement: FloatingPlacement,
  options: GetFloatingStylesOptions = {},
): CSSObject {
  const { distance = 0, offset = 0 } = options;
  const styles: CSSObject = {};

  const [position, alignment] = placement.split('-') as [Position, FloatingAlignment | undefined];

  switch (position) {
    case 'bottom': {
      styles.top = '100%';
      styles.marginTop = distance;

      if (alignment === 'start') {
        styles.left = offset;
      } else if (alignment === 'end') {
        styles.right = offset;
      } else {
        styles.left = '50%';
        styles.transform = 'translateX(-50%)';
      }

      break;
    }
    case 'top': {
      styles.bottom = '100%';
      styles.marginBottom = distance;

      if (alignment === 'start') {
        styles.left = offset;
      } else if (alignment === 'end') {
        styles.right = offset;
      } else {
        styles.left = '50%';
        styles.transform = 'translateX(-50%)';
      }

      break;
    }
    case 'left': {
      styles.right = '100%';
      styles.marginRight = distance;

      if (alignment === 'start') {
        styles.top = offset;
      } else if (alignment === 'end') {
        styles.bottom = offset;
      } else {
        styles.top = '50%';
        styles.transform = 'translateY(-50%)';
      }

      break;
    }
    case 'right': {
      styles.left = '100%';
      styles.marginLeft = distance;

      if (alignment === 'start') {
        styles.top = offset;
      } else if (alignment === 'end') {
        styles.bottom = offset;
      } else {
        styles.top = '50%';
        styles.transform = 'translateY(-50%)';
      }

      break;
    }
  }

  return styles;
}
