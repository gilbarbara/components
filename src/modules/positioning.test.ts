import { FloatingPlacement } from '~/types';

import { getComputedPlacement, getElementDimenstions, getFloatingStyles } from './positioning';

function getElements(
  triggerRect: Partial<DOMRect> = {},
  contentRect: Partial<DOMRect> = {},
  style: Partial<CSSStyleDeclaration> = {},
) {
  const triggerElement = document.createElement('div');
  const contentElement = document.createElement('div');

  triggerElement.getBoundingClientRect = () => ({
    top: 100,
    bottom: 300,
    left: 400,
    right: 500,
    width: 100,
    height: 100,
    x: 400,
    y: 200,
    ...triggerRect,
    toJSON,
  });

  contentElement.getBoundingClientRect = () => ({
    top: 0,
    bottom: 50,
    left: 0,
    right: 50,
    width: 50,
    height: 50,
    x: 0,
    y: 0,
    ...contentRect,
    toJSON,
  });

  Object.defineProperty(window, 'getComputedStyle', {
    writable: true,
    configurable: true,
    value: () => ({
      height: '100px',
      width: '100px',
      ...style,
    }),
  });

  return { triggerElement, contentElement };
}

function toJSON(this: DOMRect) {
  return this;
}

describe('getComputedPlacement', () => {
  it('should return "bottom" when triggerElement is null and placement is "auto"', () => {
    const contentElement = document.createElement('div');

    expect(getComputedPlacement('auto', null, contentElement)).toBe('bottom');
  });

  it('should return the provided placement when there is no overflow', () => {
    const { contentElement, triggerElement } = getElements();

    expect(getComputedPlacement('top', triggerElement, contentElement)).toBe('top');
  });

  it('should return a fallback placement when the requested placement overflows', () => {
    const { contentElement, triggerElement } = getElements({ top: 20, bottom: 70 });

    expect(getComputedPlacement('top-start', triggerElement, contentElement)).toBe('bottom-start');
  });

  it('should compute an auto placement based on available space', () => {
    const { contentElement, triggerElement } = getElements();

    expect(getComputedPlacement('auto', triggerElement, contentElement)).toBe('right');
  });

  it('should appends alignment if fallback has no dash but original did', () => {
    const { contentElement, triggerElement } = getElements({ left: 0 });

    expect(getComputedPlacement('left', triggerElement, contentElement)).toBe('top-start');
  });

  it('should use the fallback order when the "top" placement overflows', () => {
    // Place trigger near the top so "top" is impossible
    const { contentElement, triggerElement } = getElements({ top: 0 });

    expect(getComputedPlacement('top', triggerElement, contentElement)).toBe('bottom');
  });

  it('should use the fallback order when the "bottom" placement overflows', () => {
    // Place trigger near the top so "top" is impossible
    const { contentElement, triggerElement } = getElements({ bottom: 800 });

    expect(getComputedPlacement('bottom', triggerElement, contentElement)).toBe('top');
  });

  it('shoud returns "bottom" if no side can fit the content (bestPlacement is null)', () => {
    const { contentElement, triggerElement } = getElements(
      { top: 100, bottom: 150 },
      {},
      { width: '200px', height: '200px' },
    );

    expect(getComputedPlacement('top', triggerElement, contentElement)).toBe('bottom');
  });

  it('should handle wide content that would overflow horizontally with top placement', () => {
    const { contentElement, triggerElement } = getElements(
      { left: 900 },
      {},
      { width: '300px', height: '50px' },
    );

    expect(getComputedPlacement('top', triggerElement, contentElement)).toBe('top-end');
  });

  it('should handle tall content that would overflow vertically with left placement', () => {
    const { contentElement, triggerElement } = getElements(
      { top: 700, height: 50 },
      {},
      { height: '300px' },
    );

    expect(getComputedPlacement('left-start', triggerElement, contentElement)).toBe('left-end');
  });

  it('should handle alignment with top-start without available space', () => {
    const { contentElement, triggerElement } = getElements({ top: 20 }, {}, { width: '150px' });

    // Should not use top-start since it would overflow to the left
    expect(getComputedPlacement('top-start', triggerElement, contentElement)).toBe('bottom-start');
  });

  it('should handle right placement with extremely wide content', () => {
    const { contentElement, triggerElement } = getElements(
      { top: 200, bottom: 300, left: 800, right: 900 },
      {},
      { width: '500px', height: '50px' },
    );

    // Right placement with wide content near right edge should overflow
    expect(getComputedPlacement('right', triggerElement, contentElement)).toBe('bottom-end');
  });

  it('should handle extremely tall trigger elements', () => {
    const { contentElement, triggerElement } = getElements(
      { top: 50, bottom: 700, left: 400, right: 500, height: 650 },
      {},
      { width: '100px', height: '100px' },
    );

    expect(getComputedPlacement('auto', triggerElement, contentElement)).toBe('right');
  });

  it('should preserve alignment when possible with fallback position', () => {
    const { contentElement, triggerElement } = getElements(
      { top: 0, bottom: 50, left: 400, right: 500 },
      {},
      { width: '100px', height: '100px' },
    );

    // Top-start can't work (no space above), but bottom-start should
    expect(getComputedPlacement('top-start', triggerElement, contentElement)).toBe('bottom-start');
  });

  it('should handle corner cases (element in bottom-right corner)', () => {
    const { contentElement, triggerElement } = getElements(
      { top: 700, bottom: 750, left: 950, right: 1000 },
      {},
      { width: '100px', height: '100px' },
    );

    expect(getComputedPlacement('auto', triggerElement, contentElement)).toBe('left-end');
  });

  it('should prefer placements with less overflow when all options overflow', () => {
    const { contentElement, triggerElement } = getElements(
      { top: 150, bottom: 100, left: 150, right: 500 },
      {},
      { width: '200px', height: '200px' },
    );

    expect(getComputedPlacement('auto', triggerElement, contentElement)).toBe('bottom');
  });

  it('should respect preferredPlacements option', () => {
    const { contentElement, triggerElement } = getElements({
      top: 50,
    });

    expect(
      getComputedPlacement('top', triggerElement, contentElement, {
        preferredPlacements: ['right', 'left', 'bottom', 'top'],
      }),
    ).toBe('right');
  });
});

describe('getElementDimenstions', () => {
  it('should return the element dimensions from its bounding rect', () => {
    const element = document.createElement('div');

    // Override the element's getBoundingClientRect to simulate specific dimensions.
    element.getBoundingClientRect = () => ({
      top: 10,
      bottom: 60,
      left: 15,
      right: 65,
      width: 50,
      height: 50,
      x: 15,
      y: 10,
      toJSON,
    });

    Object.defineProperty(window, 'getComputedStyle', {
      writable: true,
      configurable: true,
      value: () => ({
        height: '40px',
        width: '80px',
      }),
    });

    expect(getElementDimenstions(element)).toMatchSnapshot();
  });
});

describe('getFloatingStyles', () => {
  it.each<FloatingPlacement>([
    'auto',
    'top',
    'top-start',
    'top-end',
    'bottom',
    'bottom-start',
    'bottom-end',
    'left',
    'left-start',
    'left-end',
    'right',
    'right-start',
    'right-end',
  ])('should return correct styles for %s placement', placement => {
    const styles = getFloatingStyles(placement, { offset: 20, distance: 10 });

    expect(styles).toMatchSnapshot();
  });
});
