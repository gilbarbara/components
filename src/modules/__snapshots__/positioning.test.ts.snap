// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getElementDimenstions > should return the element dimensions from its bounding rect 1`] = `
{
  "height": 40,
  "width": 80,
}
`;

exports[`getFloatingStyles > should return correct styles for auto placement 1`] = `{}`;

exports[`getFloatingStyles > should return correct styles for bottom placement 1`] = `
{
  "left": "50%",
  "marginTop": 10,
  "top": "100%",
  "transform": "translateX(-50%)",
}
`;

exports[`getFloatingStyles > should return correct styles for bottom-end placement 1`] = `
{
  "marginTop": 10,
  "right": 20,
  "top": "100%",
}
`;

exports[`getFloatingStyles > should return correct styles for bottom-start placement 1`] = `
{
  "left": 20,
  "marginTop": 10,
  "top": "100%",
}
`;

exports[`getFloatingStyles > should return correct styles for left placement 1`] = `
{
  "marginRight": 10,
  "right": "100%",
  "top": "50%",
  "transform": "translateY(-50%)",
}
`;

exports[`getFloatingStyles > should return correct styles for left-end placement 1`] = `
{
  "bottom": 20,
  "marginRight": 10,
  "right": "100%",
}
`;

exports[`getFloatingStyles > should return correct styles for left-start placement 1`] = `
{
  "marginRight": 10,
  "right": "100%",
  "top": 20,
}
`;

exports[`getFloatingStyles > should return correct styles for right placement 1`] = `
{
  "left": "100%",
  "marginLeft": 10,
  "top": "50%",
  "transform": "translateY(-50%)",
}
`;

exports[`getFloatingStyles > should return correct styles for right-end placement 1`] = `
{
  "bottom": 20,
  "left": "100%",
  "marginLeft": 10,
}
`;

exports[`getFloatingStyles > should return correct styles for right-start placement 1`] = `
{
  "left": "100%",
  "marginLeft": 10,
  "top": 20,
}
`;

exports[`getFloatingStyles > should return correct styles for top placement 1`] = `
{
  "bottom": "100%",
  "left": "50%",
  "marginBottom": 10,
  "transform": "translateX(-50%)",
}
`;

exports[`getFloatingStyles > should return correct styles for top-end placement 1`] = `
{
  "bottom": "100%",
  "marginBottom": 10,
  "right": 20,
}
`;

exports[`getFloatingStyles > should return correct styles for top-start placement 1`] = `
{
  "bottom": "100%",
  "left": 20,
  "marginBottom": 10,
}
`;
